src/test/groovy/org/prebid/server/functional/tests/privacy/DsaSpec.groovy

```diff
@@ -0,0 +1,424 @@
+        def ampStoredRequest = BidRequest.defaultBidRequest.tap {
+            regs.ext.dsa = dsa
+            setAccountId(ampRequest.account)
+        }
+
+        and: "Save storedRequest into DB"
+        def storedRequest = StoredRequest.getStoredRequest(ampRequest, ampStoredRequest)
+        storedRequestDao.save(storedRequest)
+
+        and: "Default bidder response without DSA"
+        def bidResponse = BidResponse.getDefaultBidResponse(ampStoredRequest).tap {
+            seatbid[0].bid[0].ext = new BidExt(dsa: null)
+        }
+
+        and: "Set bidder response"
+        bidder.setResponse(ampStoredRequest.id, bidResponse)
+
+        when: "PBS processes amp request"
+        def response = privacyPbsService.sendAmpRequest(ampRequest)
+
+        then: "PBS should reject bid"
+        assert !response.targeting
+
+        and: "Response should contain an error"
+        def bidId = bidResponse.seatbid[0].bid[0].id
+        assert response.ext?.errors[GENERIC]*.code == [5]
+        assert response.ext?.errors[GENERIC]*.message == ["BidId `$bidId` validation messages: Error: Bid \"$bidId\" missing DSA"]
+
+        where:
+        dsaRequired << [DsaRequired.REQUIRED,
+                        DsaRequired.REQUIRED_PUBLISHER_IS_ONLINE_PLATFORM]
+    }
+
+    def "Auction request should always forward DSA to bidders"() {
+        given: "Default bid request with DSA"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            regs.ext.dsa = dsa
+        }
+
+        when: "PBS processes auction request"
+        privacyPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Bidder request should contain DSA"
+        assert bidder.getBidderRequest(bidRequest.id).regs?.ext?.dsa == dsa
+
+        where:
+        dsa << [null,
+                new RequestDsa(),
+                RequestDsa.getDefaultDsa(DsaRequired.NOT_REQUIRED),
+                RequestDsa.getDefaultDsa(DsaRequired.SUPPORTED),
+                RequestDsa.getDefaultDsa(DsaRequired.REQUIRED),
+                RequestDsa.getDefaultDsa(DsaRequired.REQUIRED_PUBLISHER_IS_ONLINE_PLATFORM)]
+    }
+
+    def "Auction request should always accept bids with DSA"() {
+        given: "Default bid request with DSA"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            regs.ext.dsa = dsa
+        }
+
+        and: "Default bidder response with DSA"
+        def bidDsa = BidDsa.defaultDsa
+        def bidResponse = BidResponse.getDefaultBidResponse(bidRequest).tap {
+            seatbid[0].bid[0].ext = new BidExt(dsa: bidDsa)
+        }
+
+        and: "Set bidder response"
+        bidder.setResponse(bidRequest.id, bidResponse)
+
+        when: "PBS processes auction request"
+        def response = privacyPbsService.sendAuctionRequest(bidRequest)
+
+        then: "PBS should return bid"
+        assert response.seatbid[0].bid[0]
+
+        and: "Returned bid should contain DSA"
+        assert response.seatbid[0].bid[0].ext.dsa
```

Comment:
I believe checking that the DSA in the response is equal to what was returned from the bidder is required

Reply:
Also. Nitpick, but if we are checking whether `seatbid` contains any `bid`, we can simplify it to:
`assert response.seatbid.bid`

Reply:
> I believe checking that the DSA in the response is equal to what was returned from the bidder is required
Good catch. Looks like I wiped that check by accident...